components/server/src/workspace/workspace-starter.ts

```diff
@@ -332,46 +338,123 @@ export class WorkspaceStarter {
-            needsImageBuild =
-                forceRebuild || (await this.needsImageBuild(ctx, user, workspace, instance, additionalAuth));
-            if (needsImageBuild) {
-                instance.status.conditions = {
-                    neededImageBuild: true,
-                };
+    public async start(): Promise<Disposable> {
+        this.controlAllStartingWorkspaces().catch(log.error); // Check immediately on start
+        return repeat(() => this.controlAllStartingWorkspaces(), 10 * 1000);
+    }
+
+    private async controlAllStartingWorkspaces() {
+        const span = TraceContext.startSpan("controlAllOurStartingWorkspaces");
+        const ctx = { span };
+        const instances = await this.workspaceDb.trace(ctx).findInstancesByPhase(this.MANAGED_PHASES);
+        for (const instance of instances) {
+            switch (instance.status.phase) {
+                case "preparing":
+                    this.reconcileWorkspaceStart(ctx, instance.id);
+                    return;
+
+                case "building":
+                    this.reconcileWorkspaceStart(ctx, instance.id);
+                    return;
+
+                case "pending":
+                    // !!! Note: during pending, the handover between app-cluster and ws-manager happens. !!!
+                    // This means:
+                    //  - there is a control loop on ws-manager-bridge that checks for an upper limit a instance may be in pending phase
+                    //  - it will take some time after calling ws-manager to see the first status update
+                    if (durationLongerThanSeconds(Date.parse(instance.creationTime), 30)) {
+                        // In 99.9% this is due to timing, so we need to be a bit cautious here not to spam ourselves.
+                        return;
+                    }
+
+                    // Our time has come!
+                    this.reconcileWorkspaceStart(ctx, instance.id);
+                    return;
+
+                default:
+                    return;
             }
-            span.setTag("needsImageBuild", needsImageBuild);
-        } catch (err) {
-            // if we fail to check if the workspace needs an image build (e.g. becuase the image builder is unavailable),
-            // we must properly fail the workspace instance, i.e. set its status to stopped, deal with prebuilds etc.
-            //
-            // Once we've reached actuallyStartWorkspace that function will take care of failing the instance.
-            await this.failInstanceStart(ctx, err, workspace, instance);
-            throw err;
         }
+    }
 
-        if (needsImageBuild) {
-            this.actuallyStartWorkspace({ span }, instance, workspace, user, envVars, forceRebuild).catch((err) =>
-                log.error("actuallyStartWorkspace", err),
-            );
-            return { instanceID: instance.id };
-        }
+    public reconcileWorkspaceStart(_ctx: TraceContext, instanceId: string) {
```

Comment:
Couldn't we leave the old method as is (including the arguments) and only add the mutex (plus refetching an checking the instance phase)? Would be really good if the standard request flow is not impacted too much by the additional control loop.

Reply:
What's your concern? Size of the change, or performance implication? :thinking: 
If we want to have a reconcile method, we need to decouple it. And it turns out that the current client code already supports it, so I went ahead. :+1: In the other case I have to re-implement some stuff that I did not see much sense in.

Reply:
Yes, reloading all the state from the DB on the hot path, doesn't seem good.

Reply:
The actual reconcile method wouldn't do anything else than calling the same startWorkspace function. Yes, It would need to fetch the workspace, user, etc. before hand but that is not on the hot path.
And we also keep a more explicit call signature for this method.